File: dnars/test/dnars/ForwardInferenceTest.java
package dnars;

import java.util.ArrayList;
import java.util.List;
import org.junit.Test;
import static org.junit.Assert.*;
import com.tinkerpop.blueprints.impls.tg.TinkerGraph;
import com.tinkerpop.gremlin.scala.ScalaGraph;
import dnars.base.Statement;
import dnars.gremlin.DNarsGraph;
import dnars.inference.ForwardInference;

public class ForwardInferenceTest
{
	@Test
	public void deduction_analogy()
	{
		// M -> P  
		//		S -> M	=> S -> P ded 
		//		S ~ M	=> S -> P ana
		DNarsGraph graph = DNarsGraph.wrap(ScalaGraph.wrap(new TinkerGraph()));
		try
		{
			Statement[] kb = createAndAdd(graph,
				"cat -> animal (0.82, 0.91)",
				"water -> liquid (1.0, 0.9)",
				"tiger -> cat (0.5, 0.7)",
				"developer ~ job (1.0, 0.9)",
				"feline ~ cat (0.76, 0.83)"
			);
			for (Statement st: kb)
				ForwardInference.deduction_analogy(graph, st);
			Statement[] res = {
				StatementParser.apply("tiger -> animal " + kb[0].truth().deduction(kb[2].truth())),
				StatementParser.apply("feline -> animal " + kb[0].truth().analogy(kb[4].truth(), false))
			};
			assertGraph(graph, kb, res);
		} finally
		{
			graph.shutdown();
		}
	}
	
	@Test
	public void analogy_resemblance() 
	{
		// M ~ P ::
		//		S -> M	=> S -> P ana'
		//		S ~ M	=> S ~ P res
		DNarsGraph graph = DNarsGraph.wrap(ScalaGraph.wrap(new TinkerGraph()));
		try
		{
			Statement[] kb = createAndAdd(graph,
				"developer -> job (1.0, 0.9)",
				"cat ~ feline (1.0, 0.9)",
				"tiger -> cat (1.0, 0.9)",
				"water -> liquid (1.0, 0.9)",
				"lion ~ cat (1.0, 0.9)"
			);
			for (Statement st: kb)
				ForwardInference.analogy_resemblance(graph, st);
			Statement[] res = {
				StatementParser.apply("tiger -> feline " + kb[1].truth().analogy(kb[2].truth(), true)),
				StatementParser.apply("lion ~ feline " + kb[1].truth().resemblance(kb[4].truth()))
			};
			assertGraph(graph, kb, res);
		} finally
		{
			graph.shutdown();
		}
	}
	
	@Test
	public void abduction_comparison_analogy()
	{
		// P -> M 
		//		S -> M	=> S -> P abd, S ~ P cmp
		//		S ~ M 	=> P -> S ana
		DNarsGraph graph = DNarsGraph.wrap(ScalaGraph.wrap(new TinkerGraph()));
		try
		{
			Statement[] kb = createAndAdd(graph,
				"tiger -> cat (1.0, 0.9)",
				"water -> liquid (1.0, 0.9)",
				"developer -> job (1.0, 0.9)",
				"lion -> cat (1.0, 0.9)",
				"feline ~ cat (1.0, 0.9)"
			);
			for (Statement st: kb)
				ForwardInference.abduction_comparison_analogy(graph, st);
			Statement[] res = {
				StatementParser.apply("lion -> tiger " + kb[0].truth().abduction(kb[3].truth())),
				StatementParser.apply("lion ~ tiger " + kb[0].truth().comparison(kb[3].truth())),
				StatementParser.apply("lion -> feline " + kb[0].truth().analogy(kb[4].truth(), false)),
				StatementParser.apply("tiger -> feline " + kb[0].truth().analogy(kb[4].truth(), false)),
				StatementParser.apply("tiger -> lion " + kb[3].truth().abduction(kb[0].truth())),
				StatementParser.apply("tiger ~ lion " + kb[3].truth().comparison(kb[0].truth()))
			};
			assertGraph(graph, kb, res);
		} finally
		{
			graph.shutdown();
		}
	}
	
	private Statement[] createAndAdd(DNarsGraph graph, String... statements)
	{
		Statement[] st = new Statement[statements.length];
		for (int i = 0; i < statements.length; i++)
		{
			st[i] = StatementParser.apply(statements[i]);
			graph.addStatement(st[i]);
		}
		return st;
	}
	
	private void assertGraph(DNarsGraph graph, Statement[] kb, Statement[] res)
	{
		List<Statement> all = new ArrayList<>(kb.length + res.length);
		for (Statement st: kb)
			all.add(st);
		for (Statement st: res)
			all.add(st);
		
		Statement[] gr = graph.getAllStatements();
		try
		{
			assertEquals(all.size(), gr.length);
		} catch (AssertionError e)
		{
			System.out.println("Graph statements:");
			for (Statement st: gr)
				System.out.println(st);
			throw e;
		}
		
		for (Statement st: all)
		{
			boolean found = false;
			for (int i = 0; i < gr.length && !found; i++)
				if (st.equivalent(gr[i]))
					found = true;
			assertTrue("Statement " + st + " not found.", found);
		}
	}
}

File: siebog/test/siebog/dnars/inference/ForwardInferenceTest.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding
 * copyright ownership. The ASF licenses this file to you under
 * the Apache License, Version 2.0 (the "License") you may not
 * use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied.
 *
 * See the License for the specific language governing permissions
 * and limitations under the License.
 */

package siebog.dnars.inference

import org.junit.Test
import siebog.dnars.graph.DNarsGraphFactory
import siebog.dnars.TestUtils._
import siebog.dnars.base.StatementParser
import scala.collection.mutable.ListBuffer
import siebog.dnars.base.Statement

/**
 *
 * @author <a href="mitrovic.dejan@gmail.com">Dejan Mitrovic</a>
 */
class ForwardInferenceTest {
	@Test
	def deduction_analogy: Unit = {
		// M -> P  
		//		S -> M	=> S -> P ded 
		//		S ~ M	=> S -> P ana
		val graph = DNarsGraphFactory.create(TEST_KEYSPACE, null)
		try {
			val kb = createAndAdd(graph,
				"cat -> animal (0.82, 0.91)",
				"water -> liquid (1.0, 0.9)",
				"tiger -> cat (0.5, 0.7)",
				"developer ~ job (1.0, 0.9)",
				"feline ~ cat (0.76, 0.83)")
			val derived = new ListBuffer[Statement]()
			for (st <- kb)
				derived ++= ForwardInference.deduction_analogy(graph, st)
			graph.statements.addAll(derived)
			val res = List(
				invert(kb(3)),
				invert(kb(4)),
				StatementParser("tiger -> animal " + kb(0).truth.deduction(kb(2).truth)),
				StatementParser("feline -> animal " + kb(0).truth.analogy(kb(4).truth, false)))
			assertGraph(graph, kb.toList, res)
		} finally {
			graph.shutdown
			graph.clear
		}
	}

	@Test
	def analogy_resemblance: Unit = {
		// M ~ P ::
		//		S -> M	=> S -> P ana'
		//		S ~ M	=> S ~ P res
		val graph = DNarsGraphFactory.create(TEST_KEYSPACE, null)
		try {
			val kb = createAndAdd(graph,
				"developer -> job (0.6, 0.77)",
				"cat ~ feline (0.33, 0.51)",
				"tiger -> cat (0.95, 0.83)",
				"water -> liquid (0.63, 0.72)",
				"lion ~ cat (0.85, 0.48)")
			val derived = new ListBuffer[Statement]()
			for (st <- kb)
				derived ++= ForwardInference.analogy_resemblance(graph, st)
			graph.statements.addAll(derived)
			val st = StatementParser("lion ~ feline " + kb(1).truth.resemblance(kb(4).truth))
			val res = List(
				invert(kb(1)),
				invert(kb(4)),
				StatementParser("tiger -> feline " + kb(1).truth.analogy(kb(2).truth, true)),
				StatementParser("tiger -> lion " + kb(4).truth.analogy(kb(2).truth, true)),
				st,
				invert(st))
			assertGraph(graph, kb, res)
		} finally {
			graph.shutdown
			graph.clear
		}
	}

	@Test
	def abduction_comparison_analogy: Unit = {
		// P -> M 
		//		S -> M	=> S -> P abd, S ~ P cmp
		//		S ~ M 	=> P -> S ana
		val graph = DNarsGraphFactory.create(TEST_KEYSPACE, null)
		try {
			val kb = createAndAdd(graph,
				"tiger -> cat (1.0, 0.9)",
				"water -> liquid (0.68, 0.39)",
				"developer -> job (0.93, 0.46)",
				"lion -> cat (0.43, 0.75)",
				"feline ~ cat (0.49, 0.52)")
			val derived = new ListBuffer[Statement]()
			for (st <- kb)
				derived ++= ForwardInference.abduction_comparison_analogy(graph, st)
			graph.statements.addAll(derived)
			val con1 = StatementParser("lion ~ tiger " + kb(0).truth.comparison(kb(3).truth))
			val con2 = StatementParser("tiger ~ lion " + kb(3).truth.comparison(kb(0).truth))
			val revised = StatementParser("lion ~ tiger " + con1.truth.revision(con2.truth))
			val res = List(
				invert(kb(4)),
				StatementParser("lion -> tiger " + kb(0).truth.abduction(kb(3).truth)),
				revised,
				StatementParser("tiger -> feline " + kb(0).truth.analogy(kb(4).truth, false)),
				StatementParser("tiger -> lion " + kb(3).truth.abduction(kb(0).truth)),
				invert(revised),
				StatementParser("lion -> feline " + kb(3).truth.analogy(kb(4).truth, false)))
			assertGraph(graph, kb, res)
		} finally {
			graph.shutdown
			graph.clear
		}
	}

	@Test
	def compoundExtentional: Unit = {
		val graph = DNarsGraphFactory.create(TEST_KEYSPACE, null)
		try {
			val kb = createAndAdd(graph,
				"(cat x bird) -> eat (1.0, 0.9)",
				"tiger -> cat (1.0, 0.9)")
			val derived = new ListBuffer[Statement]()
			for (st <- kb)
				derived ++= ForwardInference.deduction_analogy(graph, st)
			graph.statements.addAll(derived)

			val ded = kb(0).truth.deduction(kb(1).truth)
			val res = List(
				StatementParser("cat -> (/ eat * bird) (1.0, 0.9)"),
				StatementParser("bird -> (/ eat cat *) (1.0, 0.9)"),
				StatementParser("(tiger x bird) -> eat " + ded),
				StatementParser("tiger -> (/ eat * bird) " + ded),
				StatementParser("bird -> (/ eat tiger *) " + ded))
			assertGraph(graph, kb, res)
		} finally {
			graph.shutdown
			graph.clear
		}
	}
}